ThreadChannel#retrieveMembers throwing ParsingException

<!--
  For questions join the official JDA discord server: https://discord.gg/0hMr4ce0tIl3SLv5
-->



[download]: https://bintray.com/dv8fromtheworld/maven/JDA/_latestVersion
[guild]: https://discord.gg/0hMr4ce0tIk3pSjp
[stack overflow]: https://stackoverflow.com/questions/tagged/java

## General Troubleshooting
  
- [x] I have checked for similar issues.
- [x] I have updated to the [latest JDA version][download].
- [x] I have checked the branches or the maintainers' PRs for upcoming bug fixes.

## Bug Report

ThreadChannel#retrieveThreadMembers does not work.

This method depends on the EntityBuilder being able to construct a ThreadMember using the `member` field in the json payload response, but this is not given. JSON response from Discord:
```json
[
    {
        "id": "930264297064517663",
        "flags": 0,
        "join_timestamp": "2022-01-11T00:58:12+00:00",
        "user_id": "187979032904728576"
    },
    {
        "id": "930264297064517663",
        "flags": 1,
        "join_timestamp": "2022-01-11T03:11:27.366000+00:00",
        "muted": false,
        "mute_config": null,
        "user_id": "381877390219018250"
    }
]
```

### Expected Behavior

ThreadMembers are built using the JSON provided.

### Code Example or Reproduction Steps

```java
ThreadChannel#retrieveThreadMembers().queue()
```

### Code for JDABuilder or DefaultShardManagerBuilder Used

N/A

### Exception or Error

```
[JDA RateLimit-Worker 2] ERROR Requester - There was an unexpected error while executing a REST request
net.dv8tion.jda.api.exceptions.ParsingException: Unable to resolve value with key member to type DataObject: null
	at net.dv8tion.jda.api.utils.data.DataObject.valueError(DataObject.java:807)
	at net.dv8tion.jda.api.utils.data.DataObject.lambda$getObject$0(DataObject.java:277)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at net.dv8tion.jda.api.utils.data.DataObject.getObject(DataObject.java:277)
	at net.dv8tion.jda.internal.entities.EntityBuilder.createThreadMember(EntityBuilder.java:1210)
	at net.dv8tion.jda.internal.entities.ThreadChannelImpl.lambda$retrieveThreadMembers$3(ThreadChannelImpl.java:172)
	at net.dv8tion.jda.internal.requests.RestActionImpl.handleSuccess(RestActionImpl.java:278)
	at net.dv8tion.jda.internal.requests.RestActionImpl.handleResponse(RestActionImpl.java:268)
	at net.dv8tion.jda.api.requests.Request.handleResponse(Request.java:259)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:237)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:141)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:124)
	at net.dv8tion.jda.internal.requests.ratelimit.BotRateLimiter$Bucket.run(BotRateLimiter.java:478)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
[ForkJoinPool.commonPool-worker-19] ERROR RestAction - RestAction queue returned failure: [ErrorResponseException] -1: net.dv8tion.jda.api.exceptions.ParsingException
net.dv8tion.jda.api.exceptions.ParsingException: Unable to resolve value with key member to type DataObject: null
	at net.dv8tion.jda.api.utils.data.DataObject.valueError(DataObject.java:807)
	at net.dv8tion.jda.api.utils.data.DataObject.lambda$getObject$0(DataObject.java:277)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at net.dv8tion.jda.api.utils.data.DataObject.getObject(DataObject.java:277)
	at net.dv8tion.jda.internal.entities.EntityBuilder.createThreadMember(EntityBuilder.java:1210)
	at net.dv8tion.jda.internal.entities.ThreadChannelImpl.lambda$retrieveThreadMembers$3(ThreadChannelImpl.java:172)
	at net.dv8tion.jda.internal.requests.RestActionImpl.handleSuccess(RestActionImpl.java:278)
	at net.dv8tion.jda.internal.requests.RestActionImpl.handleResponse(RestActionImpl.java:268)
	at net.dv8tion.jda.api.requests.Request.handleResponse(Request.java:259)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:237)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:141)
	at net.dv8tion.jda.internal.requests.Requester.execute(Requester.java:124)
	at net.dv8tion.jda.internal.requests.ratelimit.BotRateLimiter$Bucket.run(BotRateLimiter.java:478)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: net.dv8tion.jda.api.exceptions.ContextException
Caused by: net.dv8tion.jda.api.exceptions.ContextException

	at net.dv8tion.jda.api.exceptions.ContextException.here(ContextException.java:54)
	at net.dv8tion.jda.api.requests.Request.<init>(Request.java:71)
	at net.dv8tion.jda.internal.requests.RestActionImpl.queue(RestActionImpl.java:197)
	at net.dv8tion.jda.api.requests.RestAction.queue(RestAction.java:573)
	at net.dv8tion.jda.api.requests.RestAction.queue(RestAction.java:539)
	at net.dv8tion.jda.api.requests.RestAction$queue.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at Bot.onMessageReceived(Bot.groovy:45)
	at net.dv8tion.jda.api.hooks.ListenerAdapter.onEvent(ListenerAdapter.java:358)
	at net.dv8tion.jda.api.hooks.InterfacedEventManager.handle(InterfacedEventManager.java:96)
	at net.dv8tion.jda.internal.hooks.EventManagerProxy.handleInternally(EventManagerProxy.java:88)
	at net.dv8tion.jda.internal.hooks.EventManagerProxy.handle(EventManagerProxy.java:70)
	at net.dv8tion.jda.internal.JDAImpl.handleEvent(JDAImpl.java:163)
	at net.dv8tion.jda.internal.handle.MessageCreateHandler.handleInternally(MessageCreateHandler.java:111)
	at net.dv8tion.jda.internal.handle.SocketHandler.handle(SocketHandler.java:36)
	at net.dv8tion.jda.internal.requests.WebSocketClient.onDispatch(WebSocketClient.java:953)
	at net.dv8tion.jda.internal.requests.WebSocketClient.onEvent(WebSocketClient.java:840)
	at net.dv8tion.jda.internal.requests.WebSocketClient.handleEvent(WebSocketClient.java:818)
	at net.dv8tion.jda.internal.requests.WebSocketClient.onBinaryMessage(WebSocketClient.java:992)
	at com.neovisionaries.ws.client.ListenerManager.callOnBinaryMessage(ListenerManager.java:385)
	at com.neovisionaries.ws.client.ReadingThread.callOnBinaryMessage(ReadingThread.java:276)
	at com.neovisionaries.ws.client.ReadingThread.handleBinaryFrame(ReadingThread.java:996)
	at com.neovisionaries.ws.client.ReadingThread.handleFrame(ReadingThread.java:755)
	at com.neovisionaries.ws.client.ReadingThread.main(ReadingThread.java:108)
	at com.neovisionaries.ws.client.ReadingThread.runMain(ReadingThread.java:64)
	at com.neovisionaries.ws.client.WebSocketThread.run(WebSocketThread.java:45)
```

--- a/src/main/java/net/dv8tion/jda/api/entities/channel/concrete/ThreadChannel.java
+++ b/src/main/java/net/dv8tion/jda/api/entities/channel/concrete/ThreadChannel.java
@@ -30,6 +30,7 @@ import net.dv8tion.jda.api.entities.channel.unions.IThreadContainerUnion;
 import net.dv8tion.jda.api.managers.channel.concrete.ThreadChannelManager;
 import net.dv8tion.jda.api.requests.RestAction;
 import net.dv8tion.jda.api.requests.restaction.CacheRestAction;
+import net.dv8tion.jda.api.requests.restaction.pagination.ThreadMemberPaginationAction;
 import net.dv8tion.jda.api.utils.MiscUtil;
 import net.dv8tion.jda.internal.utils.Checks;
 
@@ -442,13 +443,14 @@ public interface ThreadChannel extends GuildMessageChannel, IMemberContainer, IS
     /**
      * Retrieves the {@link ThreadMember ThreadMembers} of this thread.
      *
-     * <p>This requires the {@link net.dv8tion.jda.api.requests.GatewayIntent#GUILD_MEMBERS} intent to be enabled.
+     * <p>This requires the {@link net.dv8tion.jda.api.requests.GatewayIntent#GUILD_MEMBERS} intent to be enabled
+     * in the <a href="https://discord.com/developers/applications" target="_blank">Application Dashboard</a>.
      *
-     * @return a RestAction that resolves into a List of {@link ThreadMember ThreadMembers} of this thread.
+     * @return {@link ThreadMemberPaginationAction}
      */
-    //TODO-v5: docs - documentation depends on implementation cleanup.
+    @Nonnull
     @CheckReturnValue
-    RestAction<List<ThreadMember>> retrieveThreadMembers();
+    ThreadMemberPaginationAction retrieveThreadMembers();
 
     /**
      * Whether the current account is the owner of this thread.

--- a/src/main/java/net/dv8tion/jda/api/requests/restaction/pagination/ThreadMemberPaginationAction.java
+++ b/src/main/java/net/dv8tion/jda/api/requests/restaction/pagination/ThreadMemberPaginationAction.java
@@ -35,12 +35,13 @@ import javax.annotation.Nonnull;
  *
  * <p><b>Example</b><br>
  * <pre>{@code
- * // Revoke all bans from a guild with a certain reason
- * public static CompletableFuture<AtomicInteger> countMembers(ThreadChannel thread) {
+ * // Count all thread members who are bots
+ * public static CompletableFuture<AtomicInteger> countBotMembers(ThreadChannel thread) {
  *     AtomicInteger count = new AtomicInteger();
  *     ThreadMemberPaginationAction members = thread.retrieveThreadMembers();
- *     return members.forEachAsync((ban) -> {
- *         count.incrementAndGet();
+ *     return members.forEachAsync((threadMember) -> {
+ *         if (threadMember.getUser().isBot())
+ *             count.incrementAndGet();
  *         return true; // continues iterating if this returns true
  *     }).thenApply((v) -> count);
  * }

--- a/src/main/java/net/dv8tion/jda/api/requests/restaction/pagination/ThreadMemberPaginationAction.java
+++ b/src/main/java/net/dv8tion/jda/api/requests/restaction/pagination/ThreadMemberPaginationAction.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.api.requests.restaction.pagination;
+
+import net.dv8tion.jda.api.entities.ThreadMember;
+import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
+
+import javax.annotation.Nonnull;
+
+/**
+ * {@link PaginationAction} that paginates the thread members endpoint.
+ * <br>Note that this implementation is not considered thread-safe as modifications to the cache are not done
+ * with a lock. Calling methods on this class from multiple threads is not recommended.
+ *
+ * <p><b>Must provide not-null {@link ThreadChannel} to compile a valid
+ * pagination route.</b>
+ *
+ * <p><b>Limits:</b><br>
+ * Minimum - 1
+ * <br>Maximum - 100
+ *
+ * <p><b>Example</b><br>
+ * <pre>{@code
+ * // Revoke all bans from a guild with a certain reason
+ * public static CompletableFuture<AtomicInteger> countMembers(ThreadChannel thread) {
+ *     AtomicInteger count = new AtomicInteger();
+ *     ThreadMemberPaginationAction members = thread.retrieveThreadMembers();
+ *     return members.forEachAsync((ban) -> {
+ *         count.incrementAndGet();
+ *         return true; // continues iterating if this returns true
+ *     }).thenApply((v) -> count);
+ * }
+ * }</pre>
+ *
+ * @see ThreadChannel#retrieveThreadMembers()
+ */
+public interface ThreadMemberPaginationAction extends PaginationAction<ThreadMember, ThreadMemberPaginationAction>
+{
+    /**
+     * The {@link ThreadChannel} this action fetches members for.
+     *
+     * @return The channel
+     */
+    @Nonnull
+    ThreadChannel getThreadChannel();
+}

--- a/src/main/java/net/dv8tion/jda/api/utils/data/DataObject.java
+++ b/src/main/java/net/dv8tion/jda/api/utils/data/DataObject.java
@@ -752,6 +752,33 @@ public class DataObject implements SerializableData
     }
 
     /**
+     * Renames an existing field to the new name.
+     * <br>This is a shorthand to {@link #remove(String) remove} under the old key and then {@link #put(String, Object) put} under the new key.
+     *
+     * <p>If there is nothing mapped to the old key, this does nothing.
+     *
+     * @param  key
+     *         The old key
+     * @param  newKey
+     *         The new key
+     *
+     * @throws IllegalArgumentException
+     *         If null is provided
+     *
+     * @return A DataObject with the updated value
+     */
+    @Nonnull
+    public DataObject rename(@Nonnull String key, @Nonnull String newKey)
+    {
+        Checks.notNull(key, "Key");
+        Checks.notNull(newKey, "Key");
+        if (!this.data.containsKey(key))
+            return this;
+        this.data.put(newKey, this.data.remove(key));
+        return this;
+    }
+
+    /**
      * {@link java.util.Collection} of all values in this DataObject.
      *
      * @return {@link java.util.Collection} for all values

--- a/src/main/java/net/dv8tion/jda/internal/entities/channel/concrete/ThreadChannelImpl.java
+++ b/src/main/java/net/dv8tion/jda/internal/entities/channel/concrete/ThreadChannelImpl.java
@@ -202,9 +202,9 @@ public class ThreadChannelImpl extends AbstractGuildChannelImpl<ThreadChannelImp
         return new DeferredRestAction<>(jda, ThreadMember.class,
                 () -> getThreadMemberById(id),
                 () -> {
-                    Route.CompiledRoute route = Route.Channels.GET_THREAD_MEMBER.compile(getId(), Long.toUnsignedString(id));
+                    Route.CompiledRoute route = Route.Channels.GET_THREAD_MEMBER.compile(getId(), Long.toUnsignedString(id)).withQueryParams("with_member", "true");
                     return new RestActionImpl<>(jda, route, (resp, req) ->
-                        jda.getEntityBuilder().createThreadMember(getGuild(), this, resp.getObject()));
+                        jda.getEntityBuilder().createThreadMember(getGuild(), this, resp.getObject().rename("guild_member", "member")));
                 });
     }
 

--- a/src/main/java/net/dv8tion/jda/internal/entities/channel/concrete/ThreadChannelImpl.java
+++ b/src/main/java/net/dv8tion/jda/internal/entities/channel/concrete/ThreadChannelImpl.java
@@ -33,12 +33,10 @@ import net.dv8tion.jda.api.entities.channel.unions.IThreadContainerUnion;
 import net.dv8tion.jda.api.managers.channel.concrete.ThreadChannelManager;
 import net.dv8tion.jda.api.requests.RestAction;
 import net.dv8tion.jda.api.requests.restaction.CacheRestAction;
+import net.dv8tion.jda.api.requests.restaction.pagination.ThreadMemberPaginationAction;
 import net.dv8tion.jda.api.utils.TimeUtil;
 import net.dv8tion.jda.api.utils.cache.CacheView;
-import net.dv8tion.jda.api.utils.data.DataArray;
-import net.dv8tion.jda.api.utils.data.DataObject;
 import net.dv8tion.jda.internal.JDAImpl;
-import net.dv8tion.jda.internal.entities.EntityBuilder;
 import net.dv8tion.jda.internal.entities.GuildImpl;
 import net.dv8tion.jda.internal.entities.channel.middleman.AbstractGuildChannelImpl;
 import net.dv8tion.jda.internal.entities.channel.mixin.attribute.ISlowmodeChannelMixin;
@@ -47,6 +45,7 @@ import net.dv8tion.jda.internal.managers.channel.concrete.ThreadChannelManagerIm
 import net.dv8tion.jda.internal.requests.DeferredRestAction;
 import net.dv8tion.jda.internal.requests.RestActionImpl;
 import net.dv8tion.jda.internal.requests.Route;
+import net.dv8tion.jda.internal.requests.restaction.pagination.ThreadMemberPaginationActionImpl;
 import net.dv8tion.jda.internal.utils.Checks;
 import net.dv8tion.jda.internal.utils.Helpers;
 
@@ -55,7 +54,6 @@ import javax.annotation.Nullable;
 import java.time.OffsetDateTime;
 import java.util.Collections;
 import java.util.EnumSet;
-import java.util.LinkedList;
 import java.util.List;
 import java.util.stream.LongStream;
 
@@ -210,27 +208,11 @@ public class ThreadChannelImpl extends AbstractGuildChannelImpl<ThreadChannelImp
                 });
     }
 
+    @Nonnull
     @Override
-    public RestAction<List<ThreadMember>> retrieveThreadMembers()
+    public ThreadMemberPaginationAction retrieveThreadMembers()
     {
-        //TODO-threads: This interacts with GUILD_MEMBERS in some way. Need to test and document how.
-
-        Route.CompiledRoute route = Route.Channels.LIST_THREAD_MEMBERS.compile(getId());
-        return new RestActionImpl<>(getJDA(), route, (response, request) ->
-        {
-            EntityBuilder builder = api.getEntityBuilder();
-            List<ThreadMember> threadMembers = new LinkedList<>();
-            DataArray memberArr = response.getArray();
-
-            for (int i = 0; i < memberArr.length(); i++)
-            {
-                final DataObject object = memberArr.getObject(i);
-                //Very possible this is gonna break because we don't get user/member info with threadmembers
-                //TODO revisit the @Nonnull annotations in ThreadMember due to the lack of user/member info. Might be a time to introduce RestX entities?
-                threadMembers.add(builder.createThreadMember((GuildImpl) this.getGuild(), this, object));
-            }
-            return Collections.unmodifiableList(threadMembers);
-        });
+        return new ThreadMemberPaginationActionImpl(this);
     }
 
     @Override

--- a/src/main/java/net/dv8tion/jda/internal/handle/ThreadMembersUpdateHandler.java
+++ b/src/main/java/net/dv8tion/jda/internal/handle/ThreadMembersUpdateHandler.java
@@ -27,7 +27,6 @@ import net.dv8tion.jda.api.utils.data.DataArray;
 import net.dv8tion.jda.api.utils.data.DataObject;
 import net.dv8tion.jda.internal.JDAImpl;
 import net.dv8tion.jda.internal.entities.EntityBuilder;
-import net.dv8tion.jda.internal.entities.GuildImpl;
 import net.dv8tion.jda.internal.entities.channel.concrete.ThreadChannelImpl;
 import net.dv8tion.jda.internal.utils.UnlockHook;
 
@@ -85,7 +84,7 @@ public class ThreadMembersUpdateHandler extends SocketHandler
         for (int i = 0; i < addedMembersJson.length(); i++)
         {
             DataObject threadMemberJson = addedMembersJson.getObject(i);
-            ThreadMember threadMember = entityBuilder.createThreadMember((GuildImpl) thread.getGuild(), thread, threadMemberJson);
+            ThreadMember threadMember = entityBuilder.createThreadMember(thread.getGuild(), thread, threadMemberJson);
             addedThreadMembers.add(threadMember);
         }
 

--- a/src/main/java/net/dv8tion/jda/internal/requests/restaction/pagination/ThreadMemberPaginationActionImpl.java
+++ b/src/main/java/net/dv8tion/jda/internal/requests/restaction/pagination/ThreadMemberPaginationActionImpl.java
@@ -1,0 +1,104 @@
+/*
+ * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.internal.requests.restaction.pagination;
+
+import net.dv8tion.jda.api.entities.ThreadMember;
+import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
+import net.dv8tion.jda.api.exceptions.ParsingException;
+import net.dv8tion.jda.api.requests.Request;
+import net.dv8tion.jda.api.requests.Response;
+import net.dv8tion.jda.api.requests.restaction.pagination.ThreadMemberPaginationAction;
+import net.dv8tion.jda.api.utils.data.DataArray;
+import net.dv8tion.jda.api.utils.data.DataObject;
+import net.dv8tion.jda.internal.entities.EntityBuilder;
+import net.dv8tion.jda.internal.entities.channel.concrete.ThreadChannelImpl;
+import net.dv8tion.jda.internal.requests.Route;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+
+public class ThreadMemberPaginationActionImpl
+    extends PaginationActionImpl<ThreadMember, ThreadMemberPaginationAction>
+    implements ThreadMemberPaginationAction
+{
+    private final ThreadChannelImpl channel;
+
+    public ThreadMemberPaginationActionImpl(ThreadChannel channel)
+    {
+        super(channel.getJDA(), Route.Channels.LIST_THREAD_MEMBERS.compile(channel.getId()).withQueryParams("with_member", "true"), 1, 100, 100);
+        this.channel = (ThreadChannelImpl) channel;
+        this.order = PaginationOrder.FORWARD;
+    }
+
+    @Nonnull
+    @Override
+    public ThreadChannel getThreadChannel()
+    {
+        return channel;
+    }
+
+    @Nonnull
+    @Override
+    public EnumSet<PaginationOrder> getSupportedOrders()
+    {
+        return EnumSet.of(getOrder());
+    }
+
+    @Override
+    protected long getKey(ThreadMember it)
+    {
+        return it.getIdLong();
+    }
+
+    @Override
+    protected void handleSuccess(Response response, Request<List<ThreadMember>> request)
+    {
+        DataArray array = response.getArray();
+        List<ThreadMember> members = new ArrayList<>(array.length());
+        EntityBuilder builder = api.getEntityBuilder();
+        for (int i = 0; i < array.length(); i++)
+        {
+            try
+            {
+                DataObject object = array.getObject(i);
+                if (object.isNull("guild_member"))
+                    continue;
+                object.put("member", object.get("guild_member"));
+                ThreadMember threadMember = builder.createThreadMember(channel.getGuild(), channel, object);
+                members.add(threadMember);
+            }
+            catch (ParsingException | NullPointerException e)
+            {
+                LOG.warn("Encountered an exception in ThreadMemberPaginationAction", e);
+            }
+        }
+
+//        if (order == PaginationOrder.BACKWARD)
+//            Collections.reverse(members);
+        if (useCache)
+            cached.addAll(members);
+
+        if (!members.isEmpty())
+        {
+            last = members.get(members.size() - 1);
+            lastKey = last.getIdLong();
+        }
+        request.onSuccess(members);
+    }
+}

--- a/src/main/java/net/dv8tion/jda/internal/requests/restaction/pagination/ThreadMemberPaginationActionImpl.java
+++ b/src/main/java/net/dv8tion/jda/internal/requests/restaction/pagination/ThreadMemberPaginationActionImpl.java
@@ -79,7 +79,7 @@ public class ThreadMemberPaginationActionImpl
                 DataObject object = array.getObject(i);
                 if (object.isNull("guild_member"))
                     continue;
-                object.put("member", object.get("guild_member"));
+                object.rename("guild_member", "member");
                 ThreadMember threadMember = builder.createThreadMember(channel.getGuild(), channel, object);
                 members.add(threadMember);
             }

